### ISSUE TITLE
Widget palette builds dynamically from site-configuration files

### ISSUE THREAD

**Issue by JonFerraiolo** [maqetta/maqetta#3504]
@jhpedemonte @peller @childsb @billreed63 @aerwin @jmockler @jinsil 

See #3503 which is the back-end coding to enable the widget palette to be built dynamically from site-configuration files.

What I'm thinking is that we have a JSON structure that lists the widget categories and widget names, and that the widget palette builds itself from that JSON structure. To illustrate a highly possible version of this just to convey the concept:

```
{
    'sections':[
        'Dojo Containers':[
            'dijit.layout.BorderContainer',
            ...
        ],
        'Dojo Controls':[
            'dijit.form.Button',
            ...
        ],
        ...
}
```

My proposal is that the above list does not have to list every single widget. If there are widgets in the widgets.json file that are not the above file, we would still add those widgets to the Widget palette at the bottom of their respective section, and any widgets listed in the above file that aren't in a widgets.json file would not appear in the Widgets palette. (Also note that any widget categories listed in the widgets.json file that are not listed in the above file would be added to the end of the list of sections in the Widget palette.)

The above sketch for the JSON format almost certainly will prove to be too simplistic for our needs. In particular, the UI folks have said that the list of widgets should change depending on the composition type. For example, for a Mobile Application, we should show the mobile widgets first, and for a Sketch, maybe the Drawing Tools should be first. As a result, the file is likely to be more complex because we will need different ordering for different mobile compositions.

Note that if we change the order of widgets in the Widgets palette when you switch between a Sketch and a Mobile Application, then the Widget palette will have to rebuild itself in response to an "editorSelected" event.

Also, we probably should use display strings such as 'Dojo Containers' and 'Dojo Controls' as the index into associative arrays. Maybe simply remove the space in the "key" and then pull the display name from a separate localization file using the "key" to look up the localization string.

**Comment by peller** [comment]
might as well use the key as is.  I guess that would be a little like getText, using the English as a key, but stripping out the space algorithmically is unnecessary as it's already a valid key in a JS hash.

For the structure, I hope we can get away from a hierarchical classification and move towards a tagging system, or one-to-many groupings.  This might be a good opportunity to do that.

**Comment by JonFerraiolo** [comment]
> we can get away from a hierarchical classification and move towards a tagging system

Ultimately at page startup time the widget palette will show a list of default categories listed in a particular order, and within each category, there will be a list of widgets shown in a particular order. Somewhere, somehow, there has to be something that the software can use to know what categories to display, what order those categories should appear, and how to order the widgets within the category. 

It might be possible in a large percentage of cases to determine category and widget ordering via an algorithm, such listing the widgets alphabetically within a given category, but that falls short of the need of some sites (e.g., IBM) where we need to rank widgets from a particular library (e.g., IBM) in front of widgets of the same type from a different library (e.g., Dojo and HTML). So, one option is that we can make the tagging system more complex so that the algorithm works in a good way most of the time.

But because of all of the different situations that might occur when customers deploy Maqetta on their own servers (with their own widget libraries), there are going to be cases where an algorithm approach won't meet needs. Therefore, even if we come up with algorithmic ways of defining the order of categories and widgets, ultimately we will need to provide an option where the site administrator will be able to explicitly define the exact list of categories (and order they should appear in the palette) and the exist list of widgets within each category (and order those widgets should apepar within each category). 

Also, we need to provide different order of things in the palette depending on the composition type.

My preference would be that we offer both a reasonably simple algorithmic approach and an option to define an explicit list of widgets for a particular category. Here is a strawman proposal for the algorithmic approach:
- there is some site-level metadata that, for each composition type (mobile, desktop, sketch), defines the ordered list of category names
- each widget has some sort of tag or property that identifies the "metatype" of widget it is (e.g., "button", "combobox", or "container")
- for each category (within each composition type), the category lists the metatypes it supports, and all widgets with any of those metatypes will appear within that category
- there is some site-level metadata that, for each composition type, defines the ordered list of libraries (e.g., AcmeCompany widgets should appear before Dojo widgets, which should appear before HTML widgets)
- the widget palette logic orders widgets first alphabetically by widget name (e.g., "Button"), and if there are multiple widgets with the same name, then secondarily sorts by preferred library order

But as I said earlier, we also need an option where for a particular composition type and a particular widget category, the site metadata can explicitly define the list of widgets to appear in that category.

**Comment by peller** [comment]
> Somewhere, somehow, there has to be something that the software can use to know what categories to display, what order those categories should appear, and how to order the widgets within the category.

How about a list of initial tags as part of the metadata?  tags could represent categories, but they wouldn't be limited to a child/parent relationship like categories.

**Comment by JonFerraiolo** [comment]
Oh, forgot to close this. Addressed by this commit: c755dd5e7d924cb9a695c62ccb54f57bc2d07011
